File: jlpt/native_ios/Sources/Data/Mappers/StudyMappers.swift
import Foundation

extension VocabularyStudyResponseDTO {
    func toDomain() -> StudyItem<Vocabulary> { StudyItem(item: item.toDomain(), status: status) }
}

extension KanjiStudyResponseDTO {
    func toDomain() -> StudyItem<Kanji> { StudyItem(item: item.toDomain(), status: status) }
}

extension GrammarStudyResponseDTO {
    func toDomain() -> StudyItem<Grammar> { StudyItem(item: item.toDomain(), status: status) }
}

extension ReviewResultResponseDTO {
    func toDomain() -> ReviewResult {
        ReviewResult(status: status, intervalDays: intervalDays, nextReviewAt: ISO8601.parse(nextReviewAt))
    }
}

extension ProgressSummaryResponseDTO {
    func toDomain() -> ProgressSummary {
        var result: [StudyItemType: [StudyStatus: Int64]] = [:]
        for (typeKey, statusCounts) in countsByItemTypeAndStatus {
            guard let itemType = StudyItemType(rawValue: typeKey) else { continue }
            var counts: [StudyStatus: Int64] = [:]
            for (statusKey, count) in statusCounts {
                guard let status = StudyStatus(rawValue: statusKey) else { continue }
                counts[status] = count
            }
            result[itemType] = counts
        }
        return ProgressSummary(countsByItemTypeAndStatus: result)
    }
}

File: jlpt/native_android/app/src/main/java/com/jlpt/android/data/mapper/StudyMappers.kt
package com.jlpt.android.data.mapper

import com.jlpt.android.data.remote.dto.GrammarStudyResponseDto
import com.jlpt.android.data.remote.dto.KanjiStudyResponseDto
import com.jlpt.android.data.remote.dto.ProgressSummaryResponseDto
import com.jlpt.android.data.remote.dto.ReviewResultResponseDto
import com.jlpt.android.data.remote.dto.VocabularyStudyResponseDto
import com.jlpt.android.domain.model.ProgressSummary
import com.jlpt.android.domain.model.ReviewResult
import com.jlpt.android.domain.model.StudyItem

fun VocabularyStudyResponseDto.toDomain() = StudyItem(item = item.toDomain(), status = status)
fun KanjiStudyResponseDto.toDomain() = StudyItem(item = item.toDomain(), status = status)
fun GrammarStudyResponseDto.toDomain() = StudyItem(item = item.toDomain(), status = status)

fun ReviewResultResponseDto.toDomain() = ReviewResult(
    status = status,
    intervalDays = intervalDays,
    nextReviewAt = nextReviewAt.toInstantOrNull()
)

fun ProgressSummaryResponseDto.toDomain() = ProgressSummary(countsByItemTypeAndStatus = countsByItemTypeAndStatus)
